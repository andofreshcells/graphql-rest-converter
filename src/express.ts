/* eslint-disable @typescript-eslint/ban-ts-comment */
import _ from 'lodash'
import express, { RequestHandler, NextFunction, IRouter, Request, Response } from 'express'
import bodyParser from 'body-parser'
import { OpenAPIV3 } from 'openapi-types'
import { parse, buildSchema, print } from 'graphql'
import OpenAPIRequestCoercer from 'openapi-request-coercer'
import OpenAPIRequestValidator from 'openapi-request-validator'
import OpenAPIResponseValidator from 'openapi-response-validator'
import { getBridgeOperations } from './graphql.js'
import { pathTemplateToExpressRoute } from './pathTemplate.js'
import {
  BridgeOperation,
  BridgeOperations,
  SchemaComponents,
  CustomOperationProps,
  CreateOpenAPIGraphQLBridgeConfig,
  CreateOpenAPISchemaConfig,
  CreateMiddlewareConfig,
} from './types.js'
import { GraphQLExecutor } from './graphQLExecutor.js'
import RequestBodyObject = OpenAPIV3.RequestBodyObject
import { InvalidResponseError } from './errors.js'
import { createOpenAPISchemaWithValidate, resolveSchemaComponents } from './utils.js'
import { transformRequest } from './multipart.js'

const middlewareToPromise =
  (middleware: RequestHandler) =>
  (req: Parameters<RequestHandler>[0], res: Parameters<RequestHandler>[1]) => {
    return new Promise<ReturnType<RequestHandler>>((resolve, reject) => {
      const next = (x: unknown) => (x ? reject(x) : resolve())
      middleware(req, res, next)
    })
  }

const promiseToHandler =
  (
    handler: (
      req: Parameters<RequestHandler>[0],
      res: Parameters<RequestHandler>[1]
    ) => Promise<ReturnType<RequestHandler>>
  ) =>
  (...args: Parameters<RequestHandler>) => {
    const result = handler(args[0], args[1])
    result.catch(args[2] as NextFunction)
  }

const jsonBodyParserPromise = middlewareToPromise(bodyParser.json())
const formBodyParserPromise = middlewareToPromise(bodyParser.urlencoded())

const addOperation = <
  T extends CustomOperationProps = CustomOperationProps,
  Req extends Request = Request,
  Res extends Response = Response
>(
  router: IRouter,
  operation: BridgeOperation<T>,
  schemaComponents: SchemaComponents,
  executor: GraphQLExecutor<Req, Res>,
  config?: CreateMiddlewareConfig<Req, Res>
) => {
  const route = pathTemplateToExpressRoute(operation.path)

  // Resolving `$ref`, at least OpenAPIRequestValidator cannot handle them properly
  const parameters_ = _.cloneDeep(operation.openAPIOperation.parameters || [])
  const requestBody_ = _.cloneDeep(operation.openAPIOperation.requestBody) as
    | RequestBodyObject
    | undefined

  resolveSchemaComponents(parameters_, schemaComponents)
  resolveSchemaComponents(requestBody_, schemaComponents)

  const requestCoercer = new OpenAPIRequestCoercer.default({
    parameters: parameters_,
    requestBody: operation.requestBodyFormData === 'FORM_DATA' ? requestBody_ : undefined,
  })

  const requestValidator =
    typeof config?.validateRequest !== 'boolean' || config.validateRequest
      ? new OpenAPIRequestValidator.default({
          parameters: parameters_,
          requestBody: requestBody_,
        })
      : undefined

  const responseValidator = config?.validateResponse
    ? new OpenAPIResponseValidator.default({
        // Type in `openapi-response-validator` seems wrong
        responses: operation.openAPIOperation.responses as any,
        components: { schemas: schemaComponents },
      })
    : undefined

  const graphqlDocument_ = print(operation.graphqlDocument)

  router[operation.httpMethod](
    route,
    promiseToHandler(async (req, res) => {
      const allRequestBodyVariables = Object.keys(operation.requestBodyVariableMap)

      if (operation.requestBodyFormData === 'JSON') {
        await jsonBodyParserPromise(req, res)
      }

      if (operation.requestBodyFormData === 'FORM_DATA') {
        await formBodyParserPromise(req, res)
      }

      const supportedContentTypes = Object.keys(
        (operation.openAPIOperation.requestBody as OpenAPIV3.RequestBodyObject)?.content || {}
      )
      if (
        ['POST', 'PUT', 'PATCH'].includes(req.method) &&
        supportedContentTypes.length > 0 &&
        !supportedContentTypes.includes(req.headers['content-type'] || '')
      ) {
        res.status(415).json({
          errors: [
            {
              message: `Only "${supportedContentTypes.join(', ')}" supported`,
            },
          ],
        })
        return
      }
      const req_ = {
        ...req,
        cookies: { ...req.cookies },
        headers: { ...req.headers },
        params: { ...req.params },
        query: { ...req.query },
        body: req.body,
      }

      // TODO: Cookies not supported
      requestCoercer.coerce(req_)

      const validationResult = requestValidator ? requestValidator.validateRequest(req_) : null

      if (validationResult) {
        res.status(400).json({ errors: validationResult.errors })
        return
      }

      let variables: Record<string, unknown> = {}
      for (const parameter of operation.openAPIOperation
        .parameters as OpenAPIV3.ParameterObject[]) {
        const variableName = operation.variableMap[parameter.name] || parameter.name
        if (parameter.in === 'path') {
          variables[variableName] = req_.params[parameter.name]
        }
        if (parameter.in === 'query') {
          variables[variableName] = req_.query[parameter.name]
        }
        if (parameter.in === 'header') {
          variables[variableName] = req_.headers[parameter.name]
        }
      }
      if (operation.requestBodyFormData === 'JSON') {
        for (const requestBodyVariable of allRequestBodyVariables) {
          variables[requestBodyVariable] =
            req_.body[operation.requestBodyVariableMap[requestBodyVariable]] ||
            (operation.requestBodyIsSingleInput ? req_.body : null)
        }
      }

      if (operation.requestBodyFormData === 'FORM_DATA') {
        for (const requestBodyVariable of allRequestBodyVariables) {
          variables[requestBodyVariable] =
            req_.body[operation.requestBodyVariableMap[requestBodyVariable]] || null
        }
      }

      if (operation.requestBodyFormData === 'MULTIPART_FORM_DATA') {
        variables = {
          ...variables,
          ...transformRequest(
            req,
            graphqlDocument_,
            allRequestBodyVariables,
            operation.requestBodyVariableMap
          ),
        }
      }

      const request = {
        document: graphqlDocument_,
        variables,
        request: req as Req,
        response: res as Res,
      }
      const result = await executor(request)

      if (config?.responseTransformer) {
        const response_ = await config.responseTransformer({
          result,
          request,
          openAPISchema: {
            operation: operation.openAPIOperation,
            method: operation.httpMethod,
            path: operation.path,
          },
        })
        if (response_) {
          const { statusCode, contentType, data } = response_
          if (contentType) {
            res.contentType(contentType)
          }
          res.status(statusCode).send(data).end()
          return
        }
      }

      // There are 2 cases of errors handled here;

      // 1. If responseValidation is enabled, it will throw in both cases
      // - If there is a response validation problem
      // - If there are any graphql errors
      if (responseValidator) {
        const responseValidationErrors = responseValidator.validateResponse(200, result.data)
        if (responseValidationErrors || result.errors) {
          throw new InvalidResponseError(
            responseValidationErrors?.message || 'GraphQL Error',
            responseValidationErrors?.errors || [],
            result.errors
          )
        }
      }
      // 2. If no response validation is enabled, we will return (possibly) both the errors and a (partial) result
      if (result.errors) {
        res.status(500).json(result).end()
        return
      }
      res.status(200).json(result.data).end()
    })
  )
}

const createExpressMiddleware = <
  T extends CustomOperationProps = CustomOperationProps,
  ThisRequest extends Request = Request,
  ThisResponse extends Response = Response
>(
  operations: BridgeOperations<T>,
  executor: GraphQLExecutor<ThisRequest, ThisResponse>,
  config?: CreateMiddlewareConfig<ThisRequest, ThisResponse>
) => {
  // TODO: Avoid depending on express directly?
  const router = express.Router()

  for (const operation of operations.operations) {
    addOperation(router, operation, operations.schemaComponents, executor, config)
  }

  return router
}

export const createOpenAPIGraphQLBridge = <
  ThisRequest extends Request = Request,
  ThisResponse extends Response = Response,
  T extends CustomOperationProps = CustomOperationProps
>(
  config: CreateOpenAPIGraphQLBridgeConfig
) => {
  const { graphqlSchema, graphqlDocument, customScalars } = config

  const graphqlSchema_ =
    typeof graphqlSchema === 'string' ? buildSchema(graphqlSchema) : graphqlSchema

  const graphqlDocument_ =
    typeof graphqlDocument === 'string' ? parse(graphqlDocument) : graphqlDocument

  const operations = getBridgeOperations<T>(graphqlSchema_, graphqlDocument_, customScalars)

  return {
    getExpressMiddleware: (
      executor: GraphQLExecutor<ThisRequest, ThisResponse>,
      config?: CreateMiddlewareConfig
    ) => createExpressMiddleware(operations, executor, config),
    getOpenAPISchema: (config: CreateOpenAPISchemaConfig<T>): OpenAPIV3.Document<T> =>
      createOpenAPISchemaWithValidate<T>(operations, config),
  }
}
