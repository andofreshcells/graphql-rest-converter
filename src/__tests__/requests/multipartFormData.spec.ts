import express from 'express'
import { buildASTSchema } from 'graphql'
import { gql } from 'graphql-tag'
import { addMocksToSchema } from '@graphql-tools/mock'
import { makeExecutableSchema } from '@graphql-tools/schema'
import { createOpenAPIGraphQLBridge } from '../../express.js'
import { OpenAPIV3 } from 'openapi-types'
import request from 'supertest'
import processRequest from 'graphql-upload/processRequest.mjs'
import GraphQLUpload from 'graphql-upload/GraphQLUpload.mjs'
import { FileUpload } from 'graphql-upload/Upload.mjs'
import { execute, parse } from 'graphql/index.js'
import assert from 'node:assert'
import { Readable } from 'node:stream'

const app = express()

const schema = buildASTSchema(gql`
  scalar Upload

  type Mutation {
    uploadAFile(id: Int!, file: Upload!): String!
    uploadAMaybeFile(id: Int!, optionalFile: Upload): Boolean!
    uploadMixedFiles(id: Int!, optionalFile: Upload, requiredFile: Upload!): [String!]!
    uploadMultipleFiles(id: String!, primaryImage: Upload!, secondaryImage: Upload!): [String!]!
  }

  type Query {
    myQuery: Boolean
  }

  schema {
    mutation: Mutation
    query: Query
  }
`)

async function streamToString(stream: Readable) {
  const chunks = []
  for await (const chunk of stream) {
    chunks.push(Buffer.from(chunk))
  }
  return Buffer.concat(chunks).toString('utf-8')
}

const gqlSchema = addMocksToSchema({
  resolvers: {
    Upload: GraphQLUpload,
    Mutation: {
      uploadAFile: async (root: unknown, { file }: { file: Promise<FileUpload> }) => {
        return await streamToString((await file).createReadStream())
      },
      uploadMixedFiles: async (
        root: unknown,
        { requiredFile }: { requiredFile: Promise<FileUpload>; optionalFile: Promise<FileUpload> }
      ) => {
        // This is an edge case, technically the spec forbids this case / did not think about it.
        // We could have a timeout on the optional file and ignore it if it was not transmitted within a certain
        // time window
        const result = await requiredFile
        return [await streamToString(result.createReadStream())]
      },
      uploadMultipleFiles: async (
        root: unknown,
        {
          primaryImage,
          secondaryImage,
        }: { primaryImage: Promise<FileUpload>; secondaryImage: Promise<FileUpload> }
      ) => {
        const firstFile = await streamToString((await primaryImage).createReadStream())
        const secondFile = await streamToString((await secondaryImage).createReadStream())
        return [firstFile, secondFile]
      },
    },
  },
  schema: makeExecutableSchema({
    typeDefs: schema,
  }),
  preserveResolvers: true,
})

const scalars: Record<string, OpenAPIV3.SchemaObject> = {
  Upload: {
    type: 'string',
    format: 'binary',
  },
}

const bridge = createOpenAPIGraphQLBridge({
  graphqlSchema: gqlSchema,
  customScalars: (scalarTypeName: string) => scalars[scalarTypeName]!,
  graphqlDocument: gql`
    mutation uploadFile($id: Int!, $file: Upload! @OABody(contentType: MULTIPART_FORM_DATA))
    @OAOperation(path: "/upload-file/{id}") {
      uploadAFile(id: $id, file: $file)
    }

    mutation uploadMultipleFiles(
      $id: String!
      $primaryImage: Upload! @OABody(contentType: MULTIPART_FORM_DATA)
      $secondaryImage: Upload! @OABody(contentType: MULTIPART_FORM_DATA, path: "secondary")
    ) @OAOperation(path: "/upload-multiple-images/{id}") {
      uploadMultipleFiles(id: $id, primaryImage: $primaryImage, secondaryImage: $secondaryImage)
    }

    mutation uploadSingleOptionalFile(
      $id: Int!
      $optionalFile: Upload @OABody(contentType: MULTIPART_FORM_DATA)
    ) @OAOperation(path: "/optional-file/{id}") {
      uploadAMaybeFile(id: $id, optionalFile: $optionalFile)
    }
    mutation uploadMixedOptionalFile(
      $id: Int!
      $requiredFile: Upload! @OABody(contentType: MULTIPART_FORM_DATA)
      $optionalFile: Upload @OABody(contentType: MULTIPART_FORM_DATA)
    ) @OAOperation(path: "/mixed-optional-file/{id}") {
      uploadMixedFiles(id: $id, requiredFile: $requiredFile, optionalFile: $optionalFile)
    }
  `,
})
app.use(
  bridge.getExpressMiddleware(async ({ request, response, document, variables }) => {
    // handle non-multipart (default) graphql requests
    if (!request.is('multipart/form-data')) {
      return Promise.resolve(
        execute({
          schema: gqlSchema,
          document: parse(document),
          variableValues: variables,
        })
      )
    }
    // Process our multipart request and make sure files resolve
    const uploadRequest = await processRequest(request, response)
    // we do not support batching requests, so we can safely assume a single request
    assert(!Array.isArray(uploadRequest))
    return Promise.resolve(
      execute({
        schema: gqlSchema,
        document: parse(uploadRequest.query),
        variableValues: {
          ...variables,
          ...(uploadRequest.variables as Record<string, unknown>),
        },
      })
    )
  })
)

describe('FormData', () => {
  it('should create the correct schema', () => {
    expect(
      bridge.getOpenAPISchema({
        baseSchema: {
          openapi: '3.0.3',
          info: {
            title: 'The API',
            version: '1.0.0',
            description: 'Description',
          },
        },
      })
    ).toMatchSnapshot()
  })
  it('should support multipart form data requests with a single file', async () => {
    const buffer = Buffer.from('some data')
    const result = await request(app)
      .post('/upload-file/10')
      .attach('file', buffer, 'custom_file_name.txt')
    expect(result.body).toMatchSnapshot()
  })

  it('should support uploading multiple files', async () => {
    const buffer1 = Buffer.from('firstImage')
    const buffer2 = Buffer.from('secondImage')

    const result = await request(app)
      .post('/upload-multiple-images/10')
      .attach('primaryImage', buffer1, 'custom_file_name.txt')
      .attach('secondary', buffer2, 'custom_file_name.txt')

    expect(result.body).toMatchSnapshot()
  })

  it('should allow single optional files', async () => {
    const result = await request(app).post('/optional-file/5').send({})
    expect(result.body).toMatchSnapshot({
      uploadAMaybeFile: expect.any(Boolean),
    })
  })

  it('should throw if optional file was transmitted as `multipart/form-data`', async () => {
    await request(app)
      .post('/optional-file/5')
      .set('content-type', 'multipart/form-data')
      .expect(500)
  })

  it('should allow mixed optional files', async () => {
    // this use case is not covered by the spec and has to be handled in the implementation
    // However we still allow it
    const buffer = Buffer.from('the required file')
    const result = await request(app)
      .post('/mixed-optional-file/5')
      .attach('requiredFile', buffer, 'required_file.txt')
    expect(result.body).toMatchSnapshot()
  })

  it('should not support arrays marked with `MULTIPART_FORM_DATA`', async () => {
    // todo..
  })
})
