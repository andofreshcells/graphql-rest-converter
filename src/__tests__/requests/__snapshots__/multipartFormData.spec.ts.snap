// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormData should allow mixed optional files 1`] = `
{
  "errors": [
    {
      "message": "Only "multipart/form-data" supported",
    },
  ],
}
`;

exports[`FormData should allow single optional files 1`] = `
{
  "uploadAMaybeFile": Any<Boolean>,
}
`;

exports[`FormData should create the correct schema 1`] = `
{
  "components": {
    "schemas": {
      "Upload": {
        "description": "The \`Upload\` scalar type represents a file upload.",
        "format": "binary",
        "nullable": true,
        "type": "string",
      },
    },
  },
  "info": {
    "description": "Description",
    "title": "The API",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": {
    "/mixed-optional-file/{id}": {
      "post": {
        "operationId": "uploadMixedOptionalFile",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "format": "int32",
              "type": "integer",
            },
            "x-graphql-variable-name": "id",
          },
        ],
        "requestBody": {
          "content": {
            "multipart/form-data": {
              "schema": {
                "properties": {
                  "optionalFile": {
                    "allOf": [
                      {
                        "$ref": "#/components/schemas/Upload",
                      },
                    ],
                    "nullable": true,
                    "type": "string",
                    "x-graphql-variable-name": "optionalFile",
                  },
                  "requiredFile": {
                    "allOf": [
                      {
                        "$ref": "#/components/schemas/Upload",
                      },
                    ],
                    "nullable": false,
                    "type": "string",
                    "x-graphql-variable-name": "requiredFile",
                  },
                },
                "required": [
                  "requiredFile",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "uploadMixedFiles": {
                      "items": {
                        "type": "string",
                      },
                      "type": "array",
                    },
                  },
                  "required": [
                    "uploadMixedFiles",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "errors": {
                      "items": {
                        "properties": {
                          "errorCode": {
                            "type": "string",
                          },
                          "location": {
                            "type": "string",
                          },
                          "message": {
                            "type": "string",
                          },
                          "path": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Invalid request",
          },
          "415": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "errors": {
                      "items": {
                        "properties": {
                          "message": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Unsupported Media Type",
          },
          "500": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "data": {
                      "description": "Branch of data that does not contain errors",
                      "nullable": true,
                      "properties": {
                        "uploadMixedFiles": {
                          "items": {
                            "type": "string",
                          },
                          "type": "array",
                        },
                      },
                      "required": [
                        "uploadMixedFiles",
                      ],
                      "type": "object",
                    },
                    "errors": {
                      "items": {
                        "properties": {
                          "locations": {
                            "items": {
                              "properties": {
                                "column": {
                                  "type": "integer",
                                },
                                "line": {
                                  "type": "integer",
                                },
                              },
                              "type": "object",
                            },
                            "type": "array",
                          },
                          "message": {
                            "type": "string",
                          },
                          "path": {
                            "items": {
                              "type": "string",
                            },
                            "type": "array",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Internal Server Error",
          },
        },
        "x-graphql-operation": "mutation uploadMixedOptionalFile($id: Int!, $requiredFile: Upload!, $optionalFile: Upload) {
  uploadMixedFiles(
    id: $id
    requiredFile: $requiredFile
    optionalFile: $optionalFile
  )
}",
      },
    },
    "/optional-file/{id}": {
      "post": {
        "operationId": "uploadSingleOptionalFile",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "format": "int32",
              "type": "integer",
            },
            "x-graphql-variable-name": "id",
          },
        ],
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "nullable": true,
                "type": "object",
              },
            },
            "multipart/form-data": {
              "schema": {
                "properties": {
                  "optionalFile": {
                    "allOf": [
                      {
                        "$ref": "#/components/schemas/Upload",
                      },
                    ],
                    "nullable": true,
                    "type": "string",
                    "x-graphql-variable-name": "optionalFile",
                  },
                },
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "uploadAMaybeFile": {
                      "type": "boolean",
                    },
                  },
                  "required": [
                    "uploadAMaybeFile",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "errors": {
                      "items": {
                        "properties": {
                          "errorCode": {
                            "type": "string",
                          },
                          "location": {
                            "type": "string",
                          },
                          "message": {
                            "type": "string",
                          },
                          "path": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Invalid request",
          },
          "415": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "errors": {
                      "items": {
                        "properties": {
                          "message": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Unsupported Media Type",
          },
          "500": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "data": {
                      "description": "Branch of data that does not contain errors",
                      "nullable": true,
                      "properties": {
                        "uploadAMaybeFile": {
                          "type": "boolean",
                        },
                      },
                      "required": [
                        "uploadAMaybeFile",
                      ],
                      "type": "object",
                    },
                    "errors": {
                      "items": {
                        "properties": {
                          "locations": {
                            "items": {
                              "properties": {
                                "column": {
                                  "type": "integer",
                                },
                                "line": {
                                  "type": "integer",
                                },
                              },
                              "type": "object",
                            },
                            "type": "array",
                          },
                          "message": {
                            "type": "string",
                          },
                          "path": {
                            "items": {
                              "type": "string",
                            },
                            "type": "array",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Internal Server Error",
          },
        },
        "x-graphql-operation": "mutation uploadSingleOptionalFile($id: Int!, $optionalFile: Upload) {
  uploadAMaybeFile(id: $id, optionalFile: $optionalFile)
}",
      },
    },
    "/upload-file/{id}": {
      "post": {
        "operationId": "uploadFile",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "format": "int32",
              "type": "integer",
            },
            "x-graphql-variable-name": "id",
          },
        ],
        "requestBody": {
          "content": {
            "multipart/form-data": {
              "schema": {
                "properties": {
                  "file": {
                    "allOf": [
                      {
                        "$ref": "#/components/schemas/Upload",
                      },
                    ],
                    "nullable": false,
                    "type": "string",
                    "x-graphql-variable-name": "file",
                  },
                },
                "required": [
                  "file",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "uploadAFile": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "uploadAFile",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "errors": {
                      "items": {
                        "properties": {
                          "errorCode": {
                            "type": "string",
                          },
                          "location": {
                            "type": "string",
                          },
                          "message": {
                            "type": "string",
                          },
                          "path": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Invalid request",
          },
          "415": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "errors": {
                      "items": {
                        "properties": {
                          "message": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Unsupported Media Type",
          },
          "500": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "data": {
                      "description": "Branch of data that does not contain errors",
                      "nullable": true,
                      "properties": {
                        "uploadAFile": {
                          "type": "string",
                        },
                      },
                      "required": [
                        "uploadAFile",
                      ],
                      "type": "object",
                    },
                    "errors": {
                      "items": {
                        "properties": {
                          "locations": {
                            "items": {
                              "properties": {
                                "column": {
                                  "type": "integer",
                                },
                                "line": {
                                  "type": "integer",
                                },
                              },
                              "type": "object",
                            },
                            "type": "array",
                          },
                          "message": {
                            "type": "string",
                          },
                          "path": {
                            "items": {
                              "type": "string",
                            },
                            "type": "array",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Internal Server Error",
          },
        },
        "x-graphql-operation": "mutation uploadFile($id: Int!, $file: Upload!) {
  uploadAFile(id: $id, file: $file)
}",
      },
    },
    "/upload-multiple-images/{id}": {
      "post": {
        "operationId": "uploadMultipleFiles",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
            "x-graphql-variable-name": "id",
          },
        ],
        "requestBody": {
          "content": {
            "multipart/form-data": {
              "schema": {
                "properties": {
                  "primaryImage": {
                    "allOf": [
                      {
                        "$ref": "#/components/schemas/Upload",
                      },
                    ],
                    "nullable": false,
                    "type": "string",
                    "x-graphql-variable-name": "primaryImage",
                  },
                  "secondary": {
                    "allOf": [
                      {
                        "$ref": "#/components/schemas/Upload",
                      },
                    ],
                    "nullable": false,
                    "type": "string",
                    "x-graphql-variable-name": "secondaryImage",
                  },
                },
                "required": [
                  "primaryImage",
                  "secondary",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "uploadMultipleFiles": {
                      "items": {
                        "type": "string",
                      },
                      "type": "array",
                    },
                  },
                  "required": [
                    "uploadMultipleFiles",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "errors": {
                      "items": {
                        "properties": {
                          "errorCode": {
                            "type": "string",
                          },
                          "location": {
                            "type": "string",
                          },
                          "message": {
                            "type": "string",
                          },
                          "path": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Invalid request",
          },
          "415": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "errors": {
                      "items": {
                        "properties": {
                          "message": {
                            "type": "string",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Unsupported Media Type",
          },
          "500": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "data": {
                      "description": "Branch of data that does not contain errors",
                      "nullable": true,
                      "properties": {
                        "uploadMultipleFiles": {
                          "items": {
                            "type": "string",
                          },
                          "type": "array",
                        },
                      },
                      "required": [
                        "uploadMultipleFiles",
                      ],
                      "type": "object",
                    },
                    "errors": {
                      "items": {
                        "properties": {
                          "locations": {
                            "items": {
                              "properties": {
                                "column": {
                                  "type": "integer",
                                },
                                "line": {
                                  "type": "integer",
                                },
                              },
                              "type": "object",
                            },
                            "type": "array",
                          },
                          "message": {
                            "type": "string",
                          },
                          "path": {
                            "items": {
                              "type": "string",
                            },
                            "type": "array",
                          },
                        },
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Internal Server Error",
          },
        },
        "x-graphql-operation": "mutation uploadMultipleFiles($id: String!, $primaryImage: Upload!, $secondaryImage: Upload!) {
  uploadMultipleFiles(
    id: $id
    primaryImage: $primaryImage
    secondaryImage: $secondaryImage
  )
}",
      },
    },
  },
}
`;

exports[`FormData should support multipart form data requests with a single file 1`] = `
{
  "errors": [
    {
      "message": "Only "multipart/form-data" supported",
    },
  ],
}
`;

exports[`FormData should support uploading multiple files 1`] = `
{
  "errors": [
    {
      "message": "Only "multipart/form-data" supported",
    },
  ],
}
`;
